### PR TITLE
[Core][Mailbox] Allow user-specified synchronization degree

### PR DESCRIPTION
**Pull request by ddmbr**
This is for issue #76. It's mainly done by adding another two parameters to
`Mailbox::send_complete`.

### CODE REVIEW

File: core/hash_ring.cpp [hunk 1]
```diff
@@ -20,9 +20,18 @@
...
+}
 
-void HashRing::remove(int worker_id) { worker_set_.erase(worker_id); }
+void HashRing::remove(int worker_id) {
+    for(int i=0; i<worker_set_.size(); i++) {
+        if(worker_set_[i] == worker_id) {
+            worker_set_.erase(worker_set_.begin()+i);
+            break;
+        }
+    }
```
Comment (Yuzhen11) [inline core/hash_ring.cpp]: Is this ok? `worker_set_.erase(std::remove(worker_set_.begin(), worker_set_.end(), worker_id), worker_set_.end());`

Comment (ddmbr) [inline core/hash_ring.cpp]: Good suggestion! I'll do it.

**Comment by ddmbr** [comment]
I'm doing some more things. Please just leave this PR here until I update it...

**Comment by kygx-legend** [comment]
Is this ready?

**Comment by ddmbr** [comment]
Maybe wait until more benchmark data are available?

**Comment by ddmbr** [comment]
In terms of performance I think it's ready now

**Comment by ddmbr** [comment]
Anymore comment?

**Comment by kygx-legend** [comment]
Could you run clang-format.py and then merge this?

**Pull request merged**